automated-testing/unit-testing/tdd_example.md

```diff
@@ -0,0 +1,218 @@
+# Test-Driven Development Example
+
+With this method, rather than writing all your tests up front, you write one test a time and then switch to write the
+system code that would make that test pass. It's important to write the bare minimum of code necessary even if it is not
+actually "correct". Once the test passes you can refactor the code to make it maybe make more sense, but again the logic
+should be simple. As you write more tests, the logic gets more and more complex, but you can continue to make the
+minimal changes to the system code with confidence because all code that was written is covered.
```

Comment:
Probably worth adding something about the "gradually increasing complexity" rule. Always write the simplest test first. Avoid writing a test for the complex core of your algorithm, and always take the next simplest secondary aspect to test next.